Add readme

mccalluc opened the pull request:
@ngehlenborg : You expressed a particular concern that we use consistent language in explaining the API. How is this?

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+The API supports a small number of basic operations which can be combined to construct more complex queries.
+The API runs at a **`base_url`**: Currently `https://cells.dev.hubmapconsortium.org/api` is available.
+
+Issuing a `POST` to `{base_url}/{output_type}/` (where `output_type` is `cell`, `organ`, `gene`, or `cluster`),
```
"output_type" is not particularly informative. Are those the entities that will be retrieved by the query?

SFD5311 replied:
Yes, the output_type is the type of entity in the set that is constructed in response to the query.

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+representing the result of the operation.
+
+Three endpoints are provided for getting more information, given a query set handle:
+- `{base_url}/count/` will return the number of matching records.
```
"Records" = entities to be retrieved = `output_type` above?

mccalluc replied:
changing "records" to "entities"

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+Three endpoints are provided for getting more information, given a query set handle:
+- `{base_url}/count/` will return the number of matching records.
+- `{base_url}/evaluation/` will quickly return a limited set of fields from the records.
```
... "evaluates the query associated with the handle and returns a predefined set of fields for the entities selected by the query"?

Is "predefined" correct? I would drop "quickly" as it doesn't provide much meaning without context.

mccalluc replied:
dropped "quickly"

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+The API runs at a **`base_url`**: Currently `https://cells.dev.hubmapconsortium.org/api` is available.
+
+Issuing a `POST` to `{base_url}/{output_type}/` (where `output_type` is `cell`, `organ`, `gene`, or `cluster`),
+with query parameters in the body, will return a **query set handle**.
```
Is there any harm in calling these "query"/"query handle" rather than "query set"/"query set" handle? Is the use of "query set" a direct mapping from the Django ORM?

SFD5311 replied:
I personally don't have a strong feeling about this.  The use of query set was actually more of a name collision with the Django ORM object.

mccalluc replied:
I will update to "query"/"query handle".

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+At this point, only some of the possible combinations of constraint type and output type have been implemented.
+This matrix will be expanded over time, but queries that are better satisfied by other APIs will not be duplicated by this API.
+
+| ouput / constraint | `cell`    | `cluster` | `dataset` | `gene`    | `organ`   | `protein` |
```
```suggestion
| output / constraint | `cell`    | `cluster` | `dataset` | `gene`    | `organ`   | `protein` |
```

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+(The Python and Javascript interfaces provide a **query set** abstraction, so you don't need to deal directly with the handle.)
+
+Issuing a `POST` to `{base_url}/{operation}/` (where `operation` is `union`, `intersection`, or `difference`),
+with query set handles provided as `key_one` and `key_two` in the body, will return a new query set handle,
```
It seems that using `handle_` instead of `key_` would be more informative (but I understand that it might be too late for that).

SFD5311 replied:
Chuck suggested this a little while ago, and I'm not opposed.  I just haven't gotten to it yet.

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+## Coverage
+
+At this point, only some of the possible combinations of constraint type and output type have been implemented.
```
Will it be possible to retrieve entities by filtering on properties of the entities themselves, such as their id or other properties (cell area, cluster name, etc.)?

SFD5311 replied:
Queries by the entities id are in the works.  Queries by other properties of the entities themselves hasn't really been on my radar to this point.

mccalluc commented:
@ngehlenborg : I think I've addressed your points that are about the text. The remaining two are work items in the code itself. Re-review, if needed, and then Sean can merge.

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+Three endpoints are provided for getting more information, given a query handle:
+- `{base_url}/count/` will return the number of matching entities.
+- `{base_url}/evaluation/` will return a limited set of fields from the entities.
```
```suggestion
- `{base_url}/evaluation/` will return a limited set of fields from the entities selected by the query.
```

ngehlenborg replied:
```suggestion
- `{base_url}/evaluation/` will return a pre-defined set of fields from the entities selected by the query.
```

ngehlenborg replied:
```suggestion
- `{base_url}/evaluation/` will return a pre-defined set of fields from the entities selected by the query in random order.
```

ngehlenborg commented on README.md:
```diff
@@ -0,0 +1,42 @@
+Three endpoints are provided for getting more information, given a query handle:
+- `{base_url}/count/` will return the number of matching entities.
+- `{base_url}/evaluation/` will return a limited set of fields from the entities.
+- `{base_url}/detailevaluation/` can sort the results, and may return more fields, but may be a slower operation.
```
@SFD5311 @mccalluc I am not very fond of the uncertainty in this statement. What does "may return more fields" mean?

SFD5311 replied:
@ngehlenborg The detail evaluation, in addition to the fields returned by the basic list evaluation, returns additional values requested by the user, i.e. expression of a specified gene for a particular cell.  If no additional values are requested, the response has the same fields as the basic list evaluation.

ngehlenborg replied:
Ok. A better description of that endpoint would be "will return a both pre-defined and user-defined fields from the entities selected by the query in a user-defined or random order."

Not sure if "random" order is correct.

Any comments about retrieval performance (quickly, slower, etc.) should maybe follow in a separate paragraph below.

mccalluc replied:
I've push a fix which moves performance details to separate paragraph. I've described the ordering as "arbitrary" rather than "random".
